You own a gym 💪🏾 for pregnant women 🤰🏾 called “OB-GYM” and you recently opened a second location. You’d like to analyze its performance, but your reporting software has given you the sales data in an awkward format.


In [78]:
import numpy as np
import pandas as pd

generator = np.random.default_rng(314)

sales = pd.DataFrame({
    'date':pd.date_range(start = '2020-01-01', periods=5).repeat(2),
    'store_id':np.tile([1,2], 5),
    'sales1':np.round(generator.normal(loc=750, scale=20, size=10), 2),
    'sales2':np.round(generator.normal(loc=650, scale=40, size=10), 2),
    'members':generator.integers(low=20, high=25, size=10)
})
sales.loc[sales.store_id == 2, 'sales1'] = np.nan
sales.loc[sales.store_id == 1, 'sales2'] = np.nan

print(sales)


        date  store_id  sales1  sales2  members
0 2020-01-01         1  737.54     NaN       22
1 2020-01-01         2     NaN  629.00       20
2 2020-01-02         1  750.75     NaN       23
3 2020-01-02         2     NaN  699.01       22
4 2020-01-03         1  750.60     NaN       20
5 2020-01-03         2     NaN  640.20       24
6 2020-01-04         1  752.65     NaN       21
7 2020-01-04         2     NaN  695.64       22
8 2020-01-05         1  747.02     NaN       20
9 2020-01-05         2     NaN  632.40       22


Reshape it into a DataFrame like this

#                    sales_1  sales_2  members_1  members_2
# date
# 2020-01-01    737.54    629.00         22         20
# 2020-01-02    750.75    699.01         23         22
# 2020-01-03    750.60    640.20         20         24
# 2020-01-04    752.65    695.64         21         22
# 2020-01-05    747.02    632.40         20         22

In [79]:
temp1 = sales.groupby(['date','store_id'])['members'].agg(['sum'])
temp2 = sales[['date','sales1']].dropna()
temp3 = sales[['date','sales2']].dropna()
temp2


date  sales1
0 2020-01-01  737.54
2 2020-01-02  750.75
4 2020-01-03  750.60
6 2020-01-04  752.65
8 2020-01-05  747.02

In [82]:
temp3['date']

1   2020-01-01
3   2020-01-02
5   2020-01-03
7   2020-01-04
9   2020-01-05
Name: date, dtype: datetime64[ns]

In [83]:
result = pd.DataFrame(
{'sales_1':temp2['sales1'].values.reshape(len(temp2)),
 'sales_2':temp3['sales2'].values.reshape(len(temp3)),
 'members_1': temp1.values.reshape(len(temp1))[::2],
 'members_2': temp1.values.reshape(len(temp1))[1::2]
}, index = temp3['date']
)
result

sales_1  sales_2  members_1  members_2
date                                              
2020-01-01   737.54   629.00         22         20
2020-01-02   750.75   699.01         23         22
2020-01-03   750.60   640.20         20         24
2020-01-04   752.65   695.64         21         22
2020-01-05   747.02   632.40         20         22